### ISSUE TITLE
Filesync fails silently and breaks watches

### ISSUE THREAD

**Issue by grayside** [phase2/rig#147]
## Problem

@illepic reports that filesync's silent failures cause developers to work for an indefinite amount of time with uncertainty if troubleshooting fixes are failing because they are wrong or because the filesync isn't carrying changes properly.

As a result, we have wasted time from troubleshooting **after** a fix has been found, as well as a growing distrust in the filesync system.

## Solution

1. Re-start the effort described in #58 to support project-specific doctor checks.
2. Separate out filesync success detection into a more distinct API method that the sync:start command can use, but a rig project doctor can also leverage.
3. Recommend developers add rig doctor and rig project doctor to `rig project up`

Next level: catching problems so developers do not have to think about it. I'm raising this idea here for context but it is probably more along the lines of an enhancement after the items above are covered.

1. Build an auto-restart mechanism into/around the unison process:

**From https://asciinema.org/a/148457**

```
'unison' cpu usage of 89.6% matches resource limit [cpu usage > 50.0%]                                                                                                    
'unison' trying to restart                                                                                                                                                
'unison' stop: '/bin/bash -c supervisorctl stop unison && rm -rf /var/run/unison.pid'                                                                                     
unison: stopped                                                                                                                                                           
'unison' stopped                                                                                                                                                          
pidfile '/var/run/unison.pid' does not exist                                                                                                                              
'unison' start: '/bin/bash -c supervisorctl start unison && supervisorctl pid unison > /var/run/unison.pid'                                                               
unison: started
```

2. When unison crashes, detect it and send a notification message

Tracking down the best way to handle the detection/restart needs more research.

**Comment by grayside** [comment]
As an alternative to building out the restart of unison process inside the container (assuming it's crashing of the server and not of the client we need to be concerned with) we could take steps to build a smarter healthcheck into our unison container image and set something up to auto-restart when it reports unhealthy. This has some answers on how we might do that: https://stackoverflow.com/questions/47088261/restarting-an-unhealthy-docker-container-based-on-healthcheck

**Comment by grayside** [comment]
Separate from this track, I'm working on a `rig project sync:check` to operate as a sort of doctor check of the unison process.

**Comment by grayside** [comment]
Collecting some research avenues:

* rig project sync defaults fs.inotify.max_user_watches to 100,000 for the docker-machine. Is this enough? Probably.
  * Do we need to increase this number inside the container? Maybe.
* When lots of files change, or a file is changed while being synced from a change, it can cause the high CPU spikes mentioned in the referenced issue. https://github.com/EugenMayer/docker-image-unison/pull/11/files demonstrates using Monit to monitor for performance and use supervisord to restart the unison server process.

**Comment by mkochendorfer** [comment]
This has happened to me and other developers countless times.  It is incredibly frustrating and wastes untold hours of time going down the wrong paths debugging things that are really just that your code changes are not making it into the container.  This is by far the highest priority issue with rig currently today.

**Comment by srjosh** [comment]
I've run into this quite frequently when working on client work; it definitely seems tied in my case to my host machine going to sleep/waking up.  It is definitely frustrating.

**Comment by grayside** [comment]
Note: This issue is now: support request, problem research, "doctor" research, and autoheal research. I will probably split this apart in the next few days. I'm breaking the "doctor" angle here to #163

**Comment by grayside** [comment]
I have converted this issue to a METABUG, please re-read the issue summary for details on what we are doing so far and what this issue should continue to be used for.

**Comment by grayside** [comment]
Further discussion with afflicted users has pointed out one of the major error cases is resume-from-sleep. Improved handling of sleep/suspend/hibernation operations may go a long way to address this problem.

**Comment by mikecrittenden** [comment]
Some of us have gotten in the habit of just assuming it's broken both when starting dev (for the day or after a break) and also whenever something unexpected happens, and running sync:start proactively before doing anything else.

**Comment by febbraro** [comment]
@mikecrittenden Does that approach of always running `sync:start` more or less alleviate any of the unison problems?

**Comment by potterme** [comment]
I don't run into this with sleep, but I do run into it when sleeping+changing-networks, such as going from office to home and back.  In my experience doing ``sync:start`` always fixes it.

This is different from the unison quitting because there are too many file changes.  Changing the max_user_watches "might" help.  This often happens when doing something that seems simple, like ``mv vendor vendor_old`` or ``rm -rf node_modules``.  Deletions seem to cause the most issues.  When doing a "mv" unison sees this as both a file deletion and a file addition.

I'm not sure I'm in favor of something trying to auto-restart unison processes, since there have been cases where I've shut down unison on purpose.  But a tool to detect a problem and notify would be useful.

Education and docs on this definitely the most useful.  Once this has happened to somebody a few times they stop going down hour-long debugging rabbit holes and start checking unison more often, so even part of "rig doctor" would be helpful.  But also helpful for devs to think more about what is happening when they do stuff like "mv vendor vendor_old" and why they might be doing that.

**Comment by mikecrittenden** [comment]
@febbraro yeah that seems to handle it for me. Typically if I see issues now it's because I just forgot to run that command. I don't usually see it crash in the middle of doing something, but I might just be lucky.
